File: Leetcode/Java/121. Best Time to Buy and Sell Stock/solution.java
// https://leetcode.com/problems/best-time-to-buy-and-sell-stock/description/

class Solution {
    public int maxProfit(int[] prices) {
        int profit = 0;
        int cheapestPrice = prices[0];

        for (int idx = 0; idx < prices.length; idx++){
            cheapestPrice = Math.min(cheapestPrice, prices[idx]);
            profit = Math.max(profit, prices[idx] - cheapestPrice);
        }

        return profit;
    }
}

File: Leetcode/Typescript/121. Best Time to Buy and Sell Stock/solution.ts
// https://leetcode.com/problems/best-time-to-buy-and-sell-stock/description/

function maxProfit(prices: number[]): number {
  let curPrice = prices[0],
    maxProfit = 0;

  for (const price of prices) {
    curPrice = Math.min(curPrice, price);
    maxProfit = Math.max(maxProfit, price - curPrice);
  }

  return maxProfit;
}
